### PR TITLE
T15105 unit tests more

### PR DESCRIPTION
**Pull request by pwithnall**
First batch of unit tests and fixes for the `Consumer` stack (`file.Consumer`, `chunks.Consumer`, etc.).

More tests to be added (see the list at the bottom of `test_file.py`), but this is a good start to get merged.

https://phabricator.endlessm.com/T15105

### CODE REVIEW

File: eos_data_distribution/ndn/base.py [hunk 1]
```diff
@@ -330,7 +330,7 @@ def removePendingInterest(self, name):
     def onTimeout(self, interest, forever=False):
         name = interest.getName()
         self._callbackCount += 1
-        self.emit('interest-timeout', interest)
+        self.emit('interest-timeout', interest, forever)
```
Comment (xaiki) [inline eos_data_distribution/ndn/base.py]: try_again is a better name than forever, but either fix it in the code or in your commit message.

File: eos_data_distribution/ndn/tests/test_file.py [hunk 2]
```diff
@@ -137,6 +142,316 @@ def test_bitmap_to_num_invalid(self):
...
+
+    @property
+    def usesGLibMainContext(self):
+        return True
+
+    def setCommandSigningInfo(self, key_chain, certificate_name):
+        raise NotImplementedError('Not implemented in MockFace yet')
+
+    def expressInterest(self, interest, on_data, on_timeout):
+        self._interests[str(interest.name)] = (interest, on_data, on_timeout)
```
Comment (xaiki) [inline eos_data_distribution/ndn/tests/test_file.py]: the 'right way' to do this is interest.getName()

File: eos_data_distribution/ndn/chunks.py [hunk 3]
```diff
@@ -219,6 +219,14 @@ def _on_data(self, o, interest, data):
             self._qualified_name = name.getPrefix(-1)
 
         seg = get_chunk_component(name).toSegment()
+
+        # Have we somehow already got this segment?
+        if self._segments[seg] == SegmentState.COMPLETE:
+            logger.debug('Ignoring data ‘%s’ as it’s already been received',
+                         name)
+            self._check_for_complete()
+            return
```
Comment (xaiki) [inline eos_data_distribution/ndn/chunks.py]: the code is good, but note that this is potentially very wrong, we should **never** get 2 segments (from the same host, but in general too) or we're being messy with our requests, or the other part is being messy with it's nacks.

File: eos_data_distribution/ndn/chunks.py [hunk 4]
```diff
@@ -159,7 +160,9 @@ def _on_complete(self):
 
     def _check_for_complete(self):
         if self._segments.count(SegmentState.COMPLETE) == len(self._segments):
-            self._on_complete()
+            if not self._emitted_complete:
+                self._emitted_complete = True
+                self._on_complete()
```
Comment (xaiki) [inline eos_data_distribution/ndn/chunks.py]: it looks like we should log the case where this happens.

File: eos_data_distribution/ndn/file.py [hunk 5]
```diff
@@ -122,7 +122,14 @@ def __init__(self, *args, **kwargs):
         except ValueError as e:
             pass
 
+    def _open_files(self):
+        raise NotImplementedError()
+
     def _save_chunk(self, n, data):
+        if self._part_fd < 0:
+            self._open_files()
```
Comment (xaiki) [inline eos_data_distribution/ndn/file.py]: _save_chunk is as hot as _on_data

File: eos_data_distribution/ndn/file.py [hunk 6]
```diff
@@ -264,6 +271,7 @@ def _on_complete(self):
 
     def _create_files(self, filename):
         # XXX this is racy
+        assert self._filename
```
Comment (xaiki) [inline eos_data_distribution/ndn/file.py]: shouldn't that be filename ?

File: eos_data_distribution/ndn/base.py (hunk 1 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/base.py]: Whoops. Fixed the commit message, and will add a follow-up patch to rename `forever` to `try_again` everywhere.

File: eos_data_distribution/ndn/tests/test_file.py (hunk 2 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/tests/test_file.py]: Why is that better than accessing it as a property (which is a bit less typing)? The documentation makes them look equivalent: http://named-data.net/doc/0.4.0/PyNDN2/pyndn.html#pyndn.interest.Interest

File: eos_data_distribution/ndn/chunks.py (hunk 3 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/chunks.py]: This is a bit of defensive programming. The alternative is to write it as `assert self._segments[seg] != SegmentState.COMPLETE`, if you’re confident enough that NFD/PyNDN will always guarantee this for us. I would suggest the approach in the patch though.

File: eos_data_distribution/ndn/chunks.py (hunk 4 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/chunks.py]: OK.

File: eos_data_distribution/ndn/file.py (hunk 5 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/file.py]: I think an integer comparison should be OK to do on that path, though?

File: eos_data_distribution/ndn/file.py (hunk 6 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/file.py]: Whoops, yes. I also meant to remove the `self._filename` assignment from there (it fits better in the callers).

File: eos_data_distribution/ndn/tests/test_file.py (hunk 2 above)
Comment (xaiki) [inline eos_data_distribution/ndn/tests/test_file.py]: we tend to follow CCL, but i guess it's ok. http://named-data.net/doc/ndn-ccl-api/

File: eos_data_distribution/ndn/chunks.py (hunk 3 above)
Comment (xaiki) [inline eos_data_distribution/ndn/chunks.py]: let's be defensive

File: eos_data_distribution/ndn/file.py (hunk 5 above)
Comment (xaiki) [inline eos_data_distribution/ndn/file.py]: should be, if it's ever a problem it'll show in traces anyway.

File: eos_data_distribution/ndn/tests/test_file.py (hunk 2 above)
Comment (pwithnall) [inline eos_data_distribution/ndn/tests/test_file.py]: OK, I’ll change it.

**Pull request merged**